src/lib/timing.c

```diff
@@ -0,0 +1,230 @@
+   | This source file is subject to version 1.01 of the Xdebug license,   |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available at through the world-wide-web at                           |
+   | https://xdebug.org/license.php                                       |
+   | If you did not receive a copy of the Xdebug license and are unable   |
+   | to obtain it through the world-wide-web, please send a note to       |
+   | derick@xdebug.org so we can mail you a copy immediately.             |
+   +----------------------------------------------------------------------+
+   | Authors: Derick Rethans <derick@xdebug.org>                          |
+   |          Michael Voříšek <mvorisek@mvorisek.cz>                      |
+   +----------------------------------------------------------------------+
+ */
+
+#include "php_xdebug.h"
+#if PHP_WIN32
+# include "win32/time.h"
+# include <versionhelpers.h>
+#else
+# include <sys/time.h>
+#endif
+#if __APPLE__
+# include <mach/mach_time.h>
+#endif
+
+#include "timing.h"
+
+#define NANOTIME_MIN_STEP 10
+
+#if PHP_WIN32
+# define WIN_NANOS_IN_TICK          100
+# define WIN_TICKS_SINCE_1601_JAN_1 116444736000000000ULL
+#endif
+
+ZEND_EXTERN_MODULE_GLOBALS(xdebug)
+
+// Windows
+#if PHP_WIN32
+static uint64_t xdebug_get_nanotime_abs(xdebug_nanotime_context *nanotime_context)
+{
+	FILETIME filetime;
+
+	if (nanotime_context->win_precise_time_func != NULL) {
+		nanotime_context->win_precise_time_func(&filetime);
+		return ((((uint64_t)filetime.dwHighDateTime << 32) + (uint64_t)filetime.dwLowDateTime) - WIN_TICKS_SINCE_1601_JAN_1)
+			* WIN_NANOS_IN_TICK;
+	}
+
+	return 0;
+}
+
+// Linux/Unix with clock_gettime
+#elif HAVE_CLOCK_GETTIME
+static uint64_t xdebug_get_nanotime_abs(xdebug_nanotime_context *nanotime_context)
+{
+   struct timespec ts;
+
+#ifdef CLOCK_MONOTONIC
+	if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0) {
+		return (uint64_t)ts.tv_sec * NANOS_IN_SEC + (uint64_t)ts.tv_nsec;
+	}
+#else
+	if (clock_gettime(CLOCK_REALTIME, &ts) == 0) {
+		return (uint64_t)ts.tv_sec * NANOS_IN_SEC + (uint64_t)ts.tv_nsec;
+	}
+#endif
+
+	return 0;
+}
+
+// Fallback if better platform specific time is not available
+#elif HAVE_GETTIMEOFDAY
+static uint64_t xdebug_get_nanotime_abs(xdebug_nanotime_context *nanotime_context)
+{
+	struct timeval tp;
+
+	if (gettimeofday(&tp, NULL) == 0) {
+		return (uint64_t)tp.tv_sec * NANOS_IN_SEC + (uint64_t)tp.tv_usec * NANOS_IN_MICROSEC;
+	}
+
+	return 0;
+}
+#endif
+
+#if PHP_WIN32
+static uint64_t xdebug_counter_and_freq_to_nanotime(uint64_t counter, uint64_t freq)
+{
+	uint32_t mul = 1, freq32;
+	uint64_t q, r;
+
+	while (freq >= (1ULL << 32)) {
+		freq /= 2;
+		mul *= 2;
+	}
+	freq32 = (uint32_t)freq;
+
+	q = counter / freq32;
+	r = counter % freq32;
+	return (q * NANOS_IN_SEC + (r * NANOS_IN_SEC) / freq32) * mul;
+}
+#endif
+
+
+// Windows Windows 7 and lower
+#if PHP_WIN32
```

Comment:
can be probably removed and do not close the previous `#endif`

Reply:
Yes, it can be. But the defines are for two different functions so I prefer to have them separated.

Reply:
fine